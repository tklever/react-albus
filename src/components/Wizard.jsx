/*
 * Copyright (c) 2017 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations under
 * the License.
 */

import React, { Component, PropTypes } from 'react';
import { createMemoryHistory } from 'history';
import fixPath from '../utils';

class Wizard extends Component {
  state = {
    step: {
      path: null,
      name: null,
    },
  }

  getChildContext() {
    return {
      wizard: {
        step: this.state.step,
        steps: this.steps,
        next: this.next,
        previous: this.props.history.goBack,
        push: this.push,
        go: this.props.history.go,
        history: this.props.history,
      },
      wizardInit: this.init,
    };
  }

  componentWillMount() {
    this.unlisten = this.props.history.listen(({ pathname }) => {
      const path = pathname.split('/').pop();
      const step = this.steps.filter(s => s.path === path)[0];
      if (step) {
        this.setState({
          step,
        });
      }
    });
  }

  componentWillUnmount() {
    this.unlisten();
  }

  init = (steps) => {
    this.steps = steps;

    if (this.props.onNext) {
      this.props.onNext({ path: null, name: null }, this.steps, this.replace);
    } else {
      this.replace();
    }
  }

  steps = [];

  get paths() {
    return this.steps.map(s => s.path);
  }

  push = (step) => {
    const nextStep = step || this.paths[this.paths.indexOf(this.state.step.path) + 1];
    this.props.history.push(fixPath(`${this.props.basename}/${nextStep}`));
  }

  replace = (step) => {
    const nextStep = step || this.paths[0];
    this.props.history.replace(fixPath(`${this.props.basename}/${nextStep}`));
  }

  next = () => {
    if (this.props.onNext) {
      this.props.onNext(this.state.step, this.steps, this.push);
    } else {
      this.push();
    }
  }

  render() {
    if (this.props.render) {
      return this.props.render(this.getChildContext().wizard);
    }
    return <div className={this.props.className}>{this.props.children}</div>;
  }
}

Wizard.propTypes = {
  basename: PropTypes.string,
  children: PropTypes.node,
  className: PropTypes.string,
  history: PropTypes.shape({
    entries: PropTypes.array,
    go: PropTypes.func,
    goBack: PropTypes.func,
    listen: PropTypes.func,
    location: PropTypes.object,
    push: PropTypes.func,
    replace: PropTypes.func,
  }),
  onNext: PropTypes.func,
  render: PropTypes.func,
};

Wizard.defaultProps = {
  basename: '',
  children: null,
  className: '',
  history: createMemoryHistory(),
  onNext: null,
  render: null,
};

Wizard.childContextTypes = {
  wizard: PropTypes.object,
  wizardInit: PropTypes.func,
};

export default Wizard;
