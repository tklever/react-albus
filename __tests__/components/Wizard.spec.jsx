/*
 * Copyright (c) 2017 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations under
 * the License.
 */

import React from 'react';
import { mount } from 'enzyme';
import * as history from 'history';

import Wizard from '../../src';

history.createHistory = history.createMemoryHistory;

describe('Wizard', () => {
  let setSteps;
  let step;
  let next;
  let previous;
  let push;
  let go;
  let location;

  describe('without render prop', () => {
    it('should render without steps', () => {
      const mounted = mount(
        <Wizard />,
      );

      expect(mounted).toMatchSnapshot();
    });
  });

  describe('with no other props', () => {
    let mounted;
    const context = {
      history: {
        push: jest.fn(),
      },
    };

    beforeEach(() => {
      mounted = mount(
        <Wizard
          render={({
            _setSteps,
            step: wizardStep,
            next: wizardNext,
            previous: wizardPrevious,
            push: wizardPush,
            go: wizardGo,
          }) => {
            setSteps = _setSteps;
            step = wizardStep;
            next = wizardNext;
            previous = wizardPrevious;
            push = wizardPush;
            go = wizardGo;
            return <noscript />;
          }}
        />,
        { context },
      );

      setSteps([
        { path: 'gryffindor' },
        { path: 'slytherin' },
      ]);
    });

    it('should go to the next and previous steps', () => {
      expect(step).toEqual({ path: 'gryffindor' });
      next();
      expect(step).toEqual({ path: 'slytherin' });
      previous();
      expect(step).toEqual({ path: 'gryffindor' });
    });

    it('should push steps onto the stack', () => {
      expect(step).toEqual({ path: 'gryffindor' });
      push('slytherin');
      expect(step).toEqual({ path: 'slytherin' });
    });

    it('should pull steps off the stack', () => {
      expect(step).toEqual({ path: 'gryffindor' });
      next();
      expect(step).toEqual({ path: 'slytherin' });
      go(-1);
      expect(step).toEqual({ path: 'gryffindor' });
    });

    it('should pass history actions to it\'s parent', () => {
      push('hufflepuff');
      expect(context.history.push).toHaveBeenCalled();
    });

    it('should unlisten on unmount', () => {
      mounted.unmount();
      push('slytherin');
      expect(step.path).toEqual('gryffindor');
    });
  });

  describe('with onNext prop', () => {
    const onNext = jest.fn((onNextStep, onNextSteps, onNextPush) => onNextPush());

    beforeEach(() => {
      mount(
        <Wizard
          onNext={onNext}
          render={({
            _setSteps,
            next: wizardNext,
          }) => {
            setSteps = _setSteps;
            next = wizardNext;
            return <noscript />;
          }}
        />,
      );

      setSteps([
        { path: 'gryffindor' },
        { path: 'slytherin' },
      ]);
    });

    it('should go to the next step and call onNext', () => {
      next();
      expect(onNext).toHaveBeenCalled();
    });
  });

  describe('with routed prop', () => {
    const context = {
      history: {
        push: jest.fn(),
      },
    };

    beforeEach(() => {
      mount(
        <Wizard
          routed
          render={({
            push: wizardPush,
          }) => {
            push = wizardPush;
            return <noscript />;
          }}
        />,
        { context },
      );
    });

    it('should not pass history actions to it\'s parent', () => {
      push('hufflepuff');
      expect(context.history.push).not.toHaveBeenCalled();
    });
  });

  describe('used with a router', () => {
    it('should determine the parent path with react-router v4', () => {
      const context = {
        route: {
          match: {
            url: 'hogwarts',
          },
        },
      };

      mount(
        <Wizard
          routed
          render={({
            _setSteps,
            location: wizardLocation,
          }) => {
            setSteps = _setSteps;
            location = wizardLocation;
            return <noscript />;
          }}
        />,
        { context },
      );

      setSteps([
        { path: 'gryffindor' },
      ]);

      expect(location.pathname).toEqual('hogwarts/gryffindor');
    });

    it('should determine the parent path with react-router v3', () => {
      const context = {
        router: {
          getCurrentLocation: () => ({
            pathname: 'hogwarts',
          }),
          params: {
            splat: '/gryffindor',
          },
        },
      };

      mount(
        <Wizard
          routed
          render={({
            _setSteps,
            location: wizardLocation,
          }) => {
            setSteps = _setSteps;
            location = wizardLocation;
            return <noscript />;
          }}
        />,
        { context },
      );

      setSteps([
        { path: 'gryffindor' },
      ]);

      expect(location.pathname).toEqual('hogwarts/gryffindor');
    });
  });
});
