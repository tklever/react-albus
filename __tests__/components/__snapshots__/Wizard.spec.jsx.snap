// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Wizard without render prop should render without steps 1`] = `
<Wizard
  basename=""
  className=""
  onNext={null}
  render={null}
  routed={false}
>
  <div
    className=""
  />
</Wizard>
`;
